constellation-app/constellation
# Feature/issue 574 invalid delimited import

### Description of the Change

Handle potential errors that could occur when importing delimited files, these are grouped into 3 categories:

1. Invalid files are selected to add to source files.
2. Files that are already listed in source pane become invalid or are deleted and are selected by user in source pane.
3. Files that are already listed in source pane become invalid or are deleted and user presses import button.

The basic approahc is to try and elimiate invalid files as early as possible, but still be aware that files can change subsequent to being selected.

Rather than letting unhandled exceptions bubble up to the plugin code, catch these invalid files and use neat error dialogs that come to the front of screen, rather than being hidden under the Import Delimited File dialog.
Ie:

![WarningImport](https://user-images.githubusercontent.com/58677108/86878932-32e66000-c129-11ea-99c8-ffaf09538b75.jpg)
![SuccessfulImport](https://user-images.githubusercontent.com/58677108/86878935-34178d00-c129-11ea-9341-6c6c0bc7ac84.jpg)
![SelectMissingOrInvalidFileAlreadyInSourceList](https://user-images.githubusercontent.com/58677108/86878937-34b02380-c129-11ea-9e91-549a74d4ce79.jpg)



### Alternate Designs

Leaving code as is and allowign errors to be thrown.

### Why Should This Be In Core?

Better user experience for users importing delimited files.

### Benefits

As above

### Possible Drawbacks

None known

### Verification Process

Testing of the following:
1. Invalid files are selected to add to source files.
* Open Delimited File Importer, select JSON or XML (easy to generate invalid test files for these) press the plus button, and select one valid and one invalid file.
* confirm a warning dialog is displayed, and that after pressong OK, only the valid file is added to the list in the source pane.

2. Files that are already listed in source pane become invalid or are deleted and are selected by user in source pane.|
* Open Delimited File Importer, select JSON or XML (easy to generate invalid test files for these) press the plus button, and select two valid files.
* Confirm no errors are detected and both appear in the source file list.
* On disk delete or invalidate the contents of the file NOT being previewed in the source pane. 
* Select the now invalid (or missing) file in the source pane and confirm an warning dialog is displayed and that the file is removed from the source pane

3a. Files that are already listed in source pane become invalid or are deleted and user presses import button.
* Open Delimited File Importer, select JSON or XML (easy to generate invalid test files for these) press the plus button, and select two valid files.
* Confirm no errors are detected and both appear in the source file list.
* Map one or more fileds in the action pane.
* Invalidate or delete one of the files on disk and press the Import button.
* Confirm a warning dialog is shown and no exception dialogs thrown. 

3b. Files that are already listed in source pane become invalid or are deleted and user presses import button.
* Open Delimited File Importer, select JSON or XML (easy to generate invalid test files for these) press the plus button, and select two valid files.
* Confirm no errors are detected and both appear in the source file list.
* Map one or more fileds in the action pane.
* Confirm an info dialog is shown documenting number of rows imported. (Note rows could be duplicated, so number of data rows does not always equate to number of nodes/edges).
### Applicable Issues

https://github.com/constellation-app/constellation/issues/574
https://github.com/constellation-app/constellation/issues/606
